Add resource tracking to DXBC shader debugging

skarolewics opened the pull request:
When a resource is accessed, it is now tracked by the debug step. The shader viewer prunes resource arrays larger than 8 elements to only display accessed resources. I picked 8 as a limit arbitrarily, but it seemed like a good number where any more than that would add a decent amount of scrolling to the view.

I added another test to d3d12_resource_mapping_zoo to validate different resource access patterns.

skarolewics committed:
Add resource tracking to DXBC shader debugging

When a resource is accessed, it is now tracked by the debug step. The
shader viewer has a new panel to display resources accessed up to the
current step, with context menus to go to previous/next access of a
specific resource.

baldurk commented on qrenderdoc/Windows/ShaderViewer.cpp:
```diff
@@ -645,6 +645,81 @@ void ShaderViewer::debugShader(const ShaderBindpointMapping *bind, const ShaderR
+
+          // Walk through the states to find accessed resources
+          size_t numStates = m_States.size();
+          for(size_t s = 0; s < numStates; ++s)
```
I don't want to have any passes that iterate over the whole states list up front like this, as in future this list might be lazily populated or dynamic.

skarolewics replied:
Refactored to operate on each state and it comes in.

baldurk commented on renderdoc/api/replay/shader_types.h:
```diff
@@ -518,6 +518,34 @@ component-by-component to greatly simplify algorithms at the expense of a small
+:return: A :class:`BindpointIndex` with the binding referenced.
+:rtype: BindpointIndex
+)");
+  inline BindpointIndex GetBinding() const { return BindpointIndex(rows, columns, offset); }
```
`SourceVariableMapping` shouldn't contain any direct data like this. It should only be used to map a high-level variable (if it exists) to a debug variable, potentially adding type information that's not present in the underlying variable. Thenm the debug variable itself can then contain any contents such as a binding for a resource register. That way this will still work as-is even if no high-level variable mapping information is available.

skarolewics replied:
Switched to storing the resource accesses in a `ShaderVariableChange`

baldurk reviewed (changes requested):
I've left a couple of notes as I think this approach needs to be refactored. I didn't go in-depth since likely some of the other code will change as a result.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.cpp:
```diff
@@ -1861,6 +1861,40 @@ void FlattenVariables(const rdcstr &cbufferName, const rdcarray<ShaderConstant>
+    }
+  }
+  change.before.SetBinding(slot.registerSpace, reg, arrIdx);
+  change.after = change.before;
```
For a new register coming into scope, the `before` needs to be a default-initialised `ShaderVariable()`. That way backwards iteration knows to remove it again when stepping back over this step.

skarolewics committed:
Give ShaderViewer its own copy of bindpoint mappings

Previously this data was a reference to the current pipeline state, so
when the selected event changed, the ShaderViewer's bindpoint mappings
would be inaccurate.

skarolewics committed:
Make RichResourceTextMouseEvent only respond to left button clicks

Previously it prevented context menus from working when activated over
a rich resource text field.

skarolewics commented:
I added accessed resources to a new tab, alongside `Constants & Resources`. Currently it only shows resources as they appear during debugging, but I wrapped it in a QFrame for additional view options to be added later. I also added right-click context menus to run to previous/next access for a particular resource.

I had to make ShaderViewer store a copy of `ShaderBindpointMapping`, as it was previously referenced off the current pipeline state. This caused crashes if you changed the selected event and then stepped through an existing ShaderViewer instance, as stepping for accessed resources would access the `ShaderBindpointMapping`.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.cpp:
```diff
@@ -1861,6 +1861,70 @@ void FlattenVariables(const rdcstr &cbufferName, const rdcarray<ShaderConstant>
+  state->changes.push_back(ShaderVariableChange());
+  ShaderVariableChange &change = state->changes.back();
+  change.after.type = (type == DXBCBytecode::TYPE_RESOURCE) ? VarType::ReadOnlyResource
+                                                            : VarType::ReadWriteResource;
```
Minor nit - for various tedious reasons the `makeDebugVariableNode` doesn't give a proper preview for resource values if they don't have `rows` set to 1 since it thinks they're a struct/matrix with children elements. Would make for a nicer experience if we set `change.after.rows = change.after.columns = 1` here. E.g:

![image](https://user-images.githubusercontent.com/661798/80394964-79f2e200-88aa-11ea-84c4-9e8fc2578dc9.png)

vs

![image](https://user-images.githubusercontent.com/661798/80395051-9bec6480-88aa-11ea-9c27-715cf83ba7de.png)
